.devcontainer/devcontainer.json

```diff
@@ -0,0 +1,33 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the
+// README at: https://github.com/devcontainers/templates/tree/main/src/python
+{
+	"name": "Python 3",
+	// Or use a Dockerfile or Docker Compose file. More info: https://containers.dev/guide/dockerfile
+	"image": "mcr.microsoft.com/devcontainers/python:0-3.10",
+
+	// Features to add to the dev container. More info: https://containers.dev/features.
+	"features": {
+		"ghcr.io/devcontainers/features/python:1": {
+			"version": "none", // Use python provided by system
+			"installTools": true, // Install common Python tools like pylint
+			"installJupyterlab": true, // Install JupyterLab, a web-based interactive development environment for notebooks
+		}
+	},
+
+	// Use 'forwardPorts' to make a list of ports inside the container available locally.
+	"forwardPorts": [8501],
+	"portsAttributes": {
+		"8000": {
+			"label": "Streamlit App"
+		}
+	},
+
+	// Use 'postCreateCommand' to run commands after the container is created.
+	"postCreateCommand": "python -m pip install --user -r requirements.txt",
+
+	// Configure tool-specific properties.
+	// "customizations": {},
+
+	// Uncomment to connect as root instead. More info: https://aka.ms/dev-containers-non-root.
+	// "remoteUser": "root"
+}
```

Comment:
[Q] Is it possible to define ports mapping here?
For example, when running 
```
 streamlit run app/main.py
```
I get 

```
Collecting usage statistics. To deactivate, set browser.gatherUsageStats to False.


  You can now view your Streamlit app in your browser.

  Network URL: http://172.17.0.5:8501
  External URL: http://106.155.1.71:8501
```

But port 8501 is not mapped to the host's port.

Reply:
It doesn't need to be mapped to the host's port. VS Code will automatically forward the port for you because I've added `"forwardPorts": [8501]`. Even without adding this, when you run `make start`, VS Code will offer you to "Open in browser", which will automatically forward the `8501` port. And even if it doesn't offer that, it is possible to define custom port mappings on the VS Code port forwarding view.